stuff/cron/plagiarism_detector.py

```diff
@@ -0,0 +1,243 @@
+                                        c.`contest_id` NOT IN
+                                            (SELECT p.`contest_id` 
+                                            FROM `Plagiarisms` as p);
+                                """
+
+GET_CONTEST_SUBMISSION_IDS= """ SELECT c.contest_id, s.submission_id, s.problemset_id,
+                                s.problem_id, s.verdict, s.guid, s.language
+                                FROM Submissions as s 
+                                INNER JOIN Contests c ON c.problemset_id = s.problemset_id 
+                                WHERE c.contest_id = %s;
+                            """
+
+INSERT_INTO_PLAGIARISMS = """  
+                                INSERT INTO `Plagiarisms`
+                                (`contest_id`, `score_1`, `score_2` , `submission_id_1` , `submission_id_2`, `contents`)
+                                VALUES
+                                (%s, %s, %s, %s, %s, %s)
+                            """
+
+#Constants
+START_RED = "<span class='highlight-red'>"
+START_GREEN = "<span class='highlight-green'>"
+END = "</span>"
+
+SubmissionDownloader = Callable[[str, str], None]
+
+class S3SubmissionDownloader:
+    ''' A SubmissionDownloader that can download files from an S3 bucket.  '''
+
+    def __init__(self, bucket_name: str = 'omegaup-backup') -> None:
+        self._bucket = boto3.client('s3').Bucket(bucket_name)
+    
+    def __call__(self, guid: str, destination_path: str) -> None:
+        self._bucket.download_file(f'omegaup/submissions/{guid[:2]}/{guid[2:]}', destination_path)
+    
+class LocalSubmissionDownloader:
+
+    def __init__(self, dir: str) -> None:
+        self._dir = dir
+    
+    def __call__(self, guid: str, destination_path: str) -> None:
+        
+        shutil.copyfile(os.path.join(self._dir, f'submissions/{guid[:2]}/{guid[2:]}'), destination_path)
+
+"""
+    Function to get the range of lines that are plagiarised. 
+    The length of each list should be even. 
+    [1, 4, 7, 9] So, the ranges are 1-4 and 7-9. 
+"""
+
+def get_range(code: Sequence[str]) -> Sequence[int]:
+
+    code_range = []
+    for line_number, line in enumerate(code):
+
+        """
+            If the color is red, then it will be the same for the entire 'code'.
+            that why we are using if and not elif
+        """
+
+        if START_RED in line:
+            code_range.append(line_number)
+        elif START_GREEN in line:
+            code_range.append(line_number)
+        # We have to check for END separetly so that there is always a end for a start
+        if END in line:
+            code_range.append(line_number)
+
+    # return at one time either the Red lines or green lines range
+    return code_range
+
+def filter_and_format_result(
+        dbconn: lib.db.Connection, contest_id: int, 
+        submissions: Iterable[Mapping[str,Any]],
+        results: Sequence[Any]) -> None:
```

Comment:
```suggestion
        results: Iterable[CopyDetectorResult]) -> None:
```

we don't access `results` by index, so we can use `Iterable` (which only allows for `for x in r:`). this general idea is called [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle): always have parameters that imposes the least amount of requirements for what needs to be achieved in the function, and always return results that have the most flexibility for the caller.

Reply:
Done.